### PR TITLE
Add support for compile time string class

### PR DESCRIPTION
**Pull request by X-Ryl669**
### Description

Start implementation of a string classes for Marlin.
This first part adds compile-time string hashing methods and compile-time string splitting.

#### Compile-time string hashing
Instead of:
```cpp
if (strcmp(txt, "ABC") == 0) { }
else if (strcmp(txt, "DEC") == 0) {}
...
```
You can now write:
```cpp
switch(HASH(txt)) {
 case "ABC"_hash: break; // Here "ABC"_hash is: 23456 that's computed and replaced by the compiler
 case "DEC"_hash: break; // Here "DEC"_hash is: 1234 ...
}
```

This replace a O(N*M) algorithm with a O(N) algorithm. 

This avoids storing "ABC", "DEC" in the firmware anymore (instead only hash value are stored).

This makes code easier to read.

#### Compile-time string spliting
Instead of (pseudo code):
```cpp
   assert(__FILE__, __LINE__, ": Something is wrong");
```
You can now write:
```cpp
   assert(ONLY_FILENAME, __LINE__, ": Something is wrong");
```

That's only storing (in that example), the basename of the source file (for example: `mysource.cpp`) instead of the whole file's path when using __FILE__ (for example: `/home/me/Marlin/Marlin/src/HAL/whatever/mysource.cpp`) in the firmware.

### CODE REVIEW

**Comment by rhapsodyv** [comment]
Hum... this is a optimisation technique, but I don't think it should be used as general workflow.
I don't think it make the code easier to read, as all C/C++ programmers out there knows that the language don't support string on switch. It more confusion, as "apparently" change a language behaviour.
And we are saving what, 100 bytes?!

For me, as optimisation technique is good, as general programming is not. Maybe if we had lots of bigs string if/else, but I don't see the case here.

Anyway, programming style is up to @thinkyhead. 
@p3p is more audacious than me with C++ stuff too, he can tell his opinion.

**Comment by X-Ryl669** [comment]
Right: it's an optimization (replacing string matching algorithms from O(M*N) to O(N)).  
The gain is in not storing strings in the firmware that are only used to compare against but their hash value instead.

I've only started replacing the strcmp chains in the code, so for now, it's a small impact.
I don't expect to gain kB here but yes maybe a hundred of bytes, it's always better than nothing.

> I don't think it make the code easier to read, as all C/C++ programmers out there knows that the language don't support string on switch

Well, in that case all C++ programmers are wrong. The language now allows switching on string (since C++11) so why continue with the unefficient C strcmp way here ?

**Comment by rhapsodyv** [comment]
> Right: it's an optimization (replacing string matching algorithms from O(M*N) to O(N)).

So, optimisation technique should be used to optimize code when needed, not to obfuscate general code.

> Well, in that case all C++ programmers are wrong. The language now allows switching on string (since C++11) so why continue with the unefficient C strcmp way here ?

No it doesn't support... what it allow is doing this hash thing, thanks to `constexpr`. 

It's not because the language support doing something, that it should be used as general code. It will just be prone to maintenance issues. If not, soon we may have build our own c++ dialect based upon piles of operator overloads.

Marlin already have some macros that only a few programmers can understand, why not fill it with lots templates and operator overload to keep all non-senior c++ devs away? But it's just my humble opinion.

**Comment by X-Ryl669** [comment]
> Marlin already have some macros that only a few programmers can understand, why not fill it with lots templates and operator overload to keep all non-senior c++ devs away? But it's just my humble opinion.

I agree for the macro codes, it's very hard to grasp and honestly, I'm always referring to the macros.h file to understand what they mean (worse: for those with just a `#define Something 0` you can't use `ENABLED` or `BOTH`).

Templates can also quickly become monsters to debug so I'm limiting to the very very minimum here (not more than 1 argument, except when they are used to replace variadic macros). I've hidden the more complex template in private namespace so it does not consume developer's brain time here and does not pollute intellisense.

But operators ? Really ? It's C++ here. That's the basis of C++ to overload operators and in this case it's very useful. 

You're already using literals operators when you write "3.14f", but you're doing it so often that it's now a second nature, you don't even think of it. I think it'll be the same case for string literals (for me, it's obvious that `"something"_hash` returns the hash of a string). 

If it really bothers you, I can replace the operator by `HASH("something")`  but it seems more ugly to me.

**Comment by rhapsodyv** [comment]
My main issues are:
 - Rewriting a lot of code with almost no real gain. It’s a little more efficient, I know, but in the end it’s more a code style, if used everywhere.
 - I don’t like global operator overloading. It’s too “magic”. But it seems that _hash is pretty common. Anyway, magic code should be avoided in critical code, and marlin is critical code, in my opinion

If we go for it, I just hope that it get merged ***after the next release***.

**Comment by X-Ryl669** [comment]
Agree 100%. I don't have the right to put labels here, so please mark it as "need work" or something like this.

**Comment by thinkyhead** [comment]
We'll take a closer look at this after 2.0.8 release….

What I'm really looking for is a little helper class for fixed-size-buffer stack-based strings to wrap `strcpy_P` with `operator=`, `strcat` with `operator+=`, and so on, so that it's easier to compose little strings in-place for LCD display or serial output.

Marlin's general style is C with some borrowed C++ syntactical sugar, informal and non-pedantic. While we do gain some advantages from careful templating of classes — with the `TMCStepper` code being a good example of "configurable classes" — we should be careful not to go too far down the STL / Boost rabbit hole.

**Comment by X-Ryl669** [comment]
I started by strcmp. strcat/strcpy will come next. That won't be compile-time here, so it'll likely be simple without template.
I'm keeping an eye on the code size here. 

I'm wondering if a printf-like function would be better for some long chain of concatenation.

File: Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp [hunk 1]
```diff
@@ -1817,7 +1817,7 @@ void get_wifi_commands() {
 
         #if DISABLED(EMERGENCY_PARSER)
           // Process critical commands early
-          switch(HASH(command)) {
+          switch (HASH(command)) {
```
Comment (X-Ryl669) [inline Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: Damn it. I swear I had the extension set up correctly.

Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: The real question is, how well does this 'constexpr hash' thing work on SRAM string `command`?

File: Marlin/src/core/ctstring.h [hunk 2]
```diff
@@ -56,12 +56,10 @@ namespace CompileTimeString {
   // the string storage size in flash. So it's worth it only if the string is larger than a machine word here (likely 4 bytes)
 
   // Compute the hash of a string at compile time
-  size_t constexpr constHash(char const * input) {
-    return *input ? static_cast<size_t>(*input) + 33 * constHash( input + 1 ) : 5381;
+  size_t constexpr constHash(const char const * input) {
```
Comment (X-Ryl669) [inline Marlin/src/core/ctstring.h]: `const char * const` I think.

Comment (thinkyhead) [reply Marlin/src/core/ctstring.h]: coffee break time?

File: Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp (hunk 1 above)
Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: The constexpr *disappears* when the calling context is not compile time. So it's a basic `a * prime + accumulator` hash function here. The prime number used is 33 which has the good habit of being (32 + 1) so the multiply is replaced by a shift and an addition on most architecture (unless those with a hardware multiplier).

Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: Note that most commands (where we care about speed) start with **G** so if computing the hash is slower than comparing a single character, this will not be efficient, especially added up over the life of a print job.

Comment (rhapsodyv) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: And how about hash collisions when using the HASH thing with runtime text (user/host input, for example) ? How often we would pull our hair off with weird and hard to track bugs?

(I didn't see before that it is being used with runtime values)

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: See [here](https://godbolt.org/z/3jE137) for the compiled output of such method: TLDR: it's short and optimal.

Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: More straightforward for situations like this would be a compile-time function like this…
```cpp
if (small_strcmp(strptr, "M108")) { . . . }
```
That converts into this…
```cpp
if (strptr[0] == 'M' && strptr[1] == '1' && strptr[2] == '0' && strptr[3] == '8' && strptr[4] == '\0') { . . . }
```

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: It's not being used for G code parsing. I'm not sure it would make sense here, the codes are too short to worth it. It's being used where it matters: for matching long strings. (I've written some "when to use it" documentation in the PR, and this is exactly a case for "don't use it here").

About hash collisions: I wouldn't trust this to build a dictionary. But here we're checking a small subset of possible case (the cases in a switch is low). To have a collision with the hash function chosen, you'll need a very large string (you need to overflow the 32 bits accumulator 33 times here to collide). For such small subset, you'll hit the string buffer length limit before you succeed colliding the hash value. 

It's not like a hash table where you fold back the hash function results with a modulo into a very limited set of buckets.

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: Indeed for this emergency parsing, I could have skipped it, it does not make much sense, since the hash value takes 4 bytes and the "Mxxx" code takes 5 bytes. There's only 4 cases here, so 4 bytes to gain here, but the overhead of computing the hash will likely consume more. And we don't care if we gain few ns to avoid comparing the string four times.

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: I'll revert this part.

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: You'll need to add the "not" zero test for each char in your example. I'm not sure it's faster to have a bunch of test & branching compared to a streamlined add & shift function. 

For strcmp, memcpy and other the compiler have optimization shortcuts in its LIR representation that'll replace the code to an efficient inlined version if it's more efficient so I won't beat it here.

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: There's one case that might worth it, it's when doing progressive parsing:
```
if (str[0] == 'M') {
  if (str[1] =='1') {
    if (str[2] == '1' && str[3] == '2') {
       // M112
    } else if (str[2] == '2' && str[3] == '3') { // M123 }
  } else if (str[1] == '2' && str[3] == '4') { // M24 }
} etc...
```

Here you can skip checking for zeros and you only test a char once, but such code is a real PITA to maintain, even if it's optimal. That's what Lex and Bison are for.

Comment (rhapsodyv) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: My name "victor" generated the value `6.953.986.493.308` (using the link you shared here).... 4 letters it overflow the 32bits accumulator. I'm not sure we need 33 times this to generate a colision.

Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: The compiler should automatically rearrange a series of individual character comparisons into this result. If my constexpr `str_match` function is constructed correctly, then it should ultimately result in this exact set of comparisons.

Comment (rhapsodyv) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: > More straightforward for situations like this would be a compile-time function like this…
> 
> ```c++
> if (small_strcmp(strptr, "M108")) { . . . }
> ```
> 
> That converts into this…
> 
> ```c++
> if (strptr[0] == 'M' && strptr[1] == '1' && strptr[2] == '0' && strptr[3] == '8' && strptr[4] == '\0') { . . . }
> ```

It will only worth such optimization when used on intensive cpu operations, isn’t?
I mean, we are using about a few cycles less then using a normal strcmp...

Comment (rhapsodyv) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: int i = constHash("Yict");
 int j = constHash("zhct");

both values:
2090753662
2090753662

by hand...

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: > It will only worth such optimization when used on intensive cpu operations, isn’t?
> I mean, we are using about a few cycles less then using a normal strcmp...

You won't outsmart the compiler here. It'll always produce the best code using plain old strcmp.

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: > int i = constHash("Yict");
> int j = constHash("zhct");

You're right, you've broken all my dreams...

The value for the prime number must be higher than the highest code index we use, it should be 257 not 33. I've changed it by 33 to slow the accumulator growth factor, but it's plain wrong as you spotted.

The equation the hash function must resolve is this: `x2 + primeNumber * x1 = y2 + primeNumber * y1`
To find a collision, it means that you need to find a tuple (x2, x1) that gives the same results as (y2, y1). With a primeNumber set to 33, it's possible, you just need to use a char that's 33 slot higher than the previous (that's what you did). 

With 257, you can't have such value (since there's only 256 slots possible per char), so you shouldn't be able to find a collision... unless wrapping the accumulator.

Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: The [godbolt to prove it](https://godbolt.org/z/8xYTn6) (change to 33 to see there are multiple collision)

File: Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp [hunk 3]
```diff
@@ -1766,6 +1766,11 @@ int usartFifoAvailable(SZ_USART_FIFO *fifo) {
   return WIFISERIAL.available();
 }
 
+constexpr bool str_match(const char * const string, const char * const pattern, const int index=0) {
+  const char c = pattern[index];
+  return (string[index] == c) && (c == '\0' || str_match(string, pattern, index + 1));
+}
+
```
Comment (X-Ryl669) [inline Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: I'm almost sure that what the compiler is already doing something like this for plain old strcmp.
It's not optimizing like you expect (see [here](https://godbolt.org/z/KPKPWz), it's not converting to a chain of comparison).

File: Marlin/src/core/ctstring.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    return *str ? findStringEnd(str + 1) : str;
+  }
+
+  // Check whether a string contains a slash
+  constexpr bool containsSlash(const char * const str) {
+    return *str == '/' ? true : (*str ? containsSlash(str + 1) : false);
+  }
+  // Find the last position of the slash
+  constexpr const char* findLastSlashPos(const char * const str) {
+    return *str == '/' ? (str + 1) : findLastSlashPos(str - 1);
```
Comment (thinkyhead) [inline Marlin/src/core/ctstring.h:81]: This would be the more complete version. Should work, yes?
```cpp
    return *str == '\0' ? nullptr : (*str == '/' ? (str + 1) : findLastSlashPos(str - 1));
```

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp [hunk 5]
```diff
@@ -180,9 +181,10 @@ namespace Anycubic {
     #if ACDEBUG(AC_MARLIN)
       SERIAL_ECHOLNPAIR("ConfirmationRequest() ", msg, " printer_state:", printer_state);
     #endif
+    size_t msgH = HASH(msg);
     switch (printer_state) {
       case AC_printer_pausing: {
-        if (strcmp_P(msg, MARLIN_msg_print_paused) == 0 || strcmp_P(msg, MARLIN_msg_nozzle_parked) == 0) {
+        if (msgH == MARLIN_msg_print_paused || msgH == MARLIN_msg_nozzle_parked) {
```
Comment (thinkyhead) [inline Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp]: Actually, for this method, either it should be taking a flag / enum instead of comparing a string, or it should be simply comparing the passed string pointer to the address of the expected strings. We should not be too respectful of code from 3rd party vendors, which is often thrown together without any thoughts of optimization or future maintainability. This is one of those "meant to fix" things that just never got done.

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h [hunk 6]
```diff
@@ -87,19 +87,19 @@
...
-#define MARLIN_msg_extruder_heating    PSTR("E Heating...")
-#define MARLIN_msg_bed_heating         PSTR("Bed Heating...")
-
-#define MARLIN_msg_nozzle_parked       PSTR("Nozzle Parked")
-#define MARLIN_msg_heater_timeout      PSTR("Heater Timeout")
-#define MARLIN_msg_reheating           PSTR("Reheating...")
-#define MARLIN_msg_reheat_done         PSTR("Reheat finished.")
-#define MARLIN_msg_filament_purging    PSTR("Filament Purging...")
-#define MARLIN_msg_special_pause       PSTR("PB")
+#define MARLIN_msg_probing_failed      "Probing Failed"_hash
```
Comment (thinkyhead) [inline Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h:102]: So, again, here we see a clear shortcoming of the strategy of comparing the status message to a string (or these hashes) in order to figure out the status of the printer. The status messages are translatable strings, so they will not always be in English. Therefore the first thing to do is to get rid of this technique that the Chiron TFT is using and devise a better way for it to get the status updates it's looking for. Other displays manage to switch screens or update the text without a hackish hook into the changing of the status message. If ExtUI doesn't currently contain the "`onSomeEvent`" handler needed for the events Chiron cares about, then ExtUI can be extended.

File: Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp [hunk 7]
```diff
@@ -175,43 +175,48 @@ void AnycubicTFTClass::OnUserConfirmRequired(const char * const msg) {
...
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD Heater timeout... J18");
+        break;
+      }
+      case "Reheat finished."_hash: {
+        mediaPrintingState = AMPRINTSTATE_PAUSED;
+        mediaPauseState    = AMPAUSESTATE_REHEAT_FINISHED;
+        // enable continue button
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD Reheat done... J18");
+        break;
+      }
```
Comment (thinkyhead) [inline Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp:219]: This display is also taking a bad approach and needs the same fix.

File: Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp (hunk 3 above)
Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: The code there only has one `strcmp` instead of three with `else`. Let's see what it does with that…. And also see what `strcmp_P` looks like….

File: Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp (hunk 7 above)
Comment (rhapsodyv) [reply Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp:219]: Looks like we should do a separated PR to refactor those UI code... ?

File: Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp (hunk 3 above)
Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: Here is a clear demo of how much better the compiler's `strcmp` is compared to `str_match`:
```cpp
#include <stdio.h>
#include <stdlib.h>
#include <climits>
#include <cstring>

constexpr bool str_match(const char * const string, const char * const pattern) {
  return *string == *pattern && (*pattern == '\0' || str_match(string + 1, pattern + 1));
}

static bool check_string(const char * const txt) {
         if (str_match(txt, "M108")) { printf("match M108\n"); return true; }
    else if (str_match(txt, "M112")) { printf("match M112\n"); return true; }
    else if (str_match(txt, "M410")) { printf("match M410\n"); return true; }
    return false;
}

static bool check_string_cmp(const char * const txt) {
         if (strcmp(txt, "M108") == 0) { printf("match M108\n"); return true; }
    else if (strcmp(txt, "M112") == 0) { printf("match M112\n"); return true; }
    else if (strcmp(txt, "M410") == 0) { printf("match M410\n"); return true; }
    return false;
}

int main(int argc, char ** argv) {

    if (check_string(argv[0])) printf("String checks out with str_match.\n");

    if (check_string_cmp(argv[0])) printf("String checks out with strcmp.\n");

    return 0;
}
```

File: Marlin/src/core/ctstring.h (hunk 4 above)
Comment (X-Ryl669) [reply Marlin/src/core/ctstring.h:81]: No. The method search backward (from the end position). If you add this check it, it exits immediately. And it's only called if there is a slash in the string so it'll always finish.

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp (hunk 5 above)
Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp]: I haven't found where are the string coming from (there are defined once but not used anywhere else). 
So I've deduced they come from an external peripheral (the chiron TFT system ?).

So I've modified the define to store the hash of the string instead of the strings. They are quite big so this gives quite a saving here.

If we can modify the source code from the peripheral, then yes an enum would be better. 

In fact, the `switch { switch {` code could probably be reversed (only testing the message's hash here and skip the printer state's switch all together), since I think they've written the code this way to avoid doing too much strcmp here.

This would even provide more code saving.

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h (hunk 6 above)
Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h:102]: You're right and you know a lot more than me here. Can we make this change ? Those strings seems like they are only used here, so I don't really know where they come from.

File: Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp (hunk 7 above)
Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp:219]: Where does it come from ? The `msg` parameter, I mean.

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h (hunk 6 above)
Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h:102]: Is there a event broadcaster system in Marlin, it would simplify a lot of these codes (found in 2 UI it seems) ?

File: Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp (hunk 3 above)
Comment (X-Ryl669) [reply Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp]: The strcmp is a lot more compact (~1.5x more compact, I'd say). This is mainly due to the fact that it's not inlined compared to str_match, so in your example, you've 3 str_match instance for 1 strcmp instance.

disassembly:
```asm
_main:
0000000100000df7	pushq	%rbp
0000000100000df8	movq	%rsp, %rbp
0000000100000dfb	pushq	%r14
0000000100000dfd	pushq	%rbx
0000000100000dfe	movq	%rsi, %r14
0000000100000e01	movq	(%r14), %rbx
0000000100000e04	movl	$0x1, %ecx
0000000100000e09	movzbl	(%rbx), %eax
0000000100000e0c	cmpl	$0x4d, %eax
0000000100000e0f	jne	0x100000ea6
0000000100000e15	movb	$0x4d, %sil
// str_match section
0000000100000e18	leaq	0x121(%rip), %rdx ## literal pool for: "M108"
0000000100000e1f	movl	$0x1, %eax
0000000100000e24	testb	%sil, %sil
0000000100000e27	je	0x100000e79
0000000100000e29	movzbl	(%rbx,%rcx), %esi
0000000100000e2d	movzbl	(%rcx,%rdx), %edi
0000000100000e31	incq	%rcx
0000000100000e34	cmpl	%edi, %esi
0000000100000e36	je	0x100000e24
0000000100000e38	movb	$0x4d, %sil
0000000100000e3b	leaq	0x103(%rip), %rdx ## literal pool for: "M112"
0000000100000e42	movl	$0x1, %ecx
0000000100000e47	testb	%sil, %sil
0000000100000e4a	je	0x100000e82
0000000100000e4c	movzbl	(%rbx,%rax), %esi
0000000100000e50	movzbl	(%rax,%rdx), %edi
0000000100000e54	incq	%rax
0000000100000e57	cmpl	%edi, %esi
0000000100000e59	je	0x100000e47
0000000100000e5b	movb	$0x4d, %dl
0000000100000e5d	leaq	0xe6(%rip), %rax ## literal pool for: "M410"
0000000100000e64	testb	%dl, %dl
0000000100000e66	je	0x100000e8b
0000000100000e68	movzbl	(%rbx,%rcx), %edx
0000000100000e6c	movzbl	(%rcx,%rax), %esi
0000000100000e70	incq	%rcx
0000000100000e73	cmpl	%esi, %edx
0000000100000e75	je	0x100000e64
0000000100000e77	jmp	0x100000ea6
0000000100000e79	leaq	0x128(%rip), %rdi ## literal pool for: "match M108"
0000000100000e80	jmp	0x100000e92
0000000100000e82	leaq	0x114(%rip), %rdi ## literal pool for: "match M112"
0000000100000e89	jmp	0x100000e92
0000000100000e8b	leaq	0x100(%rip), %rdi ## literal pool for: "match M410"
0000000100000e92	callq	0x100000f10 ## symbol stub for: _puts
0000000100000e97	leaq	0xd2(%rip), %rdi ## literal pool for: "String checks out with str_match."
0000000100000e9e	callq	0x100000f10 ## symbol stub for: _puts
0000000100000ea3	movq	(%r14), %rbx
// Strcmp section
0000000100000ea6	leaq	0x93(%rip), %rsi ## literal pool for: "M108"
0000000100000ead	movq	%rbx, %rdi
0000000100000eb0	callq	0x100000f16 ## symbol stub for: _strcmp
0000000100000eb5	testl	%eax, %eax
0000000100000eb7	je	0x100000ee8
0000000100000eb9	leaq	0x85(%rip), %rsi ## literal pool for: "M112"
0000000100000ec0	movq	%rbx, %rdi
0000000100000ec3	callq	0x100000f16 ## symbol stub for: _strcmp
0000000100000ec8	testl	%eax, %eax
0000000100000eca	je	0x100000ef1
0000000100000ecc	leaq	0x77(%rip), %rsi ## literal pool for: "M410"
0000000100000ed3	movq	%rbx, %rdi
0000000100000ed6	callq	0x100000f16 ## symbol stub for: _strcmp
0000000100000edb	testl	%eax, %eax
0000000100000edd	jne	0x100000f09
0000000100000edf	leaq	0xac(%rip), %rdi ## literal pool for: "match M410"
0000000100000ee6	jmp	0x100000ef8
0000000100000ee8	leaq	0xb9(%rip), %rdi ## literal pool for: "match M108"
0000000100000eef	jmp	0x100000ef8
0000000100000ef1	leaq	0xa5(%rip), %rdi ## literal pool for: "match M112"
0000000100000ef8	callq	0x100000f10 ## symbol stub for: _puts
0000000100000efd	leaq	0x4c(%rip), %rdi ## literal pool for: "String checks out with strcmp."
0000000100000f04	callq	0x100000f10 ## symbol stub for: _puts
// End of strcmp section
0000000100000f09	xorl	%eax, %eax
0000000100000f0b	popq	%rbx
0000000100000f0c	popq	%r14
0000000100000f0e	popq	%rbp
0000000100000f0f	retq
```

**Comment by X-Ryl669** [comment]
In the end, if we rework the 2 tft code to avoid matching string, it does not make much sense to add this compile time string hashing in Marlin.  The few places left with strcmp are pretty efficient already (and the gain for the LVGL's keyboard is 16 bytes on my system, might not worth the trouble and the additional source code inflation & future maintenance)

I think it'll be better to factor out the events the 2 TFT code are using and make some kind of event broadcaster/listener so they can register on a event's enum instead of parsing strings. 

If you are ok with this, please close the PR, I'll try to replace the messy code by event handling instead.

**Comment by rhapsodyv** [comment]
I like event listeners. But they should be usage for, well, event things. With too much decoupling, the code becomes hard to follow and understand by eye.

It would be nice some sort of meeting/chat to talk about marlin architecture.

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h (hunk 6 above)
Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft_defs.h:102]: The `ExtUI` API works on a callback system, but it might not yet have all the hooks the developer was looking for. That's okay, because it is always being extended to fit different needs.

To follow up on the Chiron code I would look at each of the status messages that the Chiron is acting upon, find out where that message is being set, and make sure there is a suitable `ExtUI` hook being called at that point. If there is no hook, I would add one. Finally, I would modify the Chiron code to implement those hooks instead of examining the status message.

**Comment by thinkyhead** [comment]
The closest we've got to an 'event system' (View & Input) is an API like `ExtUI` which works on callbacks — not a bad architecture. Everything is static, and all `ExtUI` displays have the opportunity to handle messages uniquely. Events don't go to a queue, but are instantly dispatched to the display. If the display wanted to queue its events, it could do so. User interface events differ by display, and while there are some shared methods, each display type has a distinct way of turning button clicks, touches, and other events into effects.

In terms of APIs that help standardize… as part of the Feature API (posted as a Project) there is the concept of "edit/display values" which encapsulate a value, a getter, a setter, a range limit, a display style for editing, flags for scaling and acceleration, etc. An `EditValue` object that always applies the same prototype is going to be more suitable for hooking into menu edit items, especially since edit items are currently carrying this information which really belongs with the data class. It will initially apply to `MarlinUI`, but later it can also improve consistency for displays that implement a custom UI.

Small steps….

File: Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp (hunk 7 above)
Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp:219]: Various => `ExtUI::onUserConfirmRequired(STRING)` => `AnycubicTFTClass::OnUserConfirmRequired(STRING)`.

**Proposal:** Add a parameter to `onUserConfirmRequired` taking an `enum` that corresponds to each unique confirm message, so that the enum value can be compared instead of the string. Maybe the string pointer doesn't need to be passed, and functions that take the enum can call a function to get the string based on the enum. This function would then be used in other places where an enum could be passed instead of a string pointer.

Comment (thinkyhead) [reply Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp:219]: A small number of messages are ever sent to `onUserConfirmRequired_P` so it's quite doable:
```cpp
PSTR("Load Filament")
PSTR("Filament Purging...")
GET_TEXT(MSG_NOZZLE_PARKED)
GET_TEXT(MSG_HEATER_TIMEOUT)
PSTR("Reheat finished.")
PSTR("MMU2 Eject Recover")
PSTR("M43 Wait Called")
GET_TEXT(MSG_USERWAIT)
GET_TEXT(MSG_PRINT_DONE)
PSTR("Delta Calibration in progress")
PSTR("Stow Probe")
```

**Comment by sjasonsmith** [comment]
@thinkyhead, after reading through all of this I realize it terminated in the following:

> If you are ok with this, please close the PR, I'll try to replace the messy code by event handling instead.

Is there any reason to still keep this open?

**Comment by X-Ryl669** [comment]
Stale PR, and not sure it's useful, so I'm closing it.

**Comment by thinkyhead** [comment]
I hope to incorporate a little class into Marlin to make the code prettier when composing strings on the stack, with at least `operator+` implemented for C-strings, `__FlashStringHelper` strings, and perhaps Arduino `String` strings too. Compile-time strings are certainly interesting and instructive, so we may revisit this approach in future.
